Lua: Execute callback on keystroke

tjdevries opened the pull request:
Proposal:

Add ability to run lua to inspect current keys and do whatever you want with them.

Why:

I want to show recently typed keys in a little floating window for streaming :)


@kdheepak Here's what I was thinking for taking the key logging / viewer to a whole different level.

tjdevries committed:
feat: Add vim.register_keystroke_callback

tjdevries commented:
TODO:
- [ ] Some more testing
- [ ] Disable the check when you don't have anything in the table (it is very inexpensive to do, but I would still like to make it have "no change" for current users).

kdheepak commented:
Super neat! I like how simple and elegant the implementation is :)

tjdevries commented:
Oh, I forgot to mention motivation. `-w` doesn't work in real time, and it seems weird to make it work in real time as well.

You can see some related and compiled information here: https://github.com/kdheepak/keystrokes.nvim Shoutout to @kdheepak for inspiring me to think about how to do this.

kdheepak commented on src/nvim/lua/vim.lua:
```diff
@@ -478,4 +478,52 @@ function vim.defer_fn(fn, timeout)
+  end
+
+  if not vim.tbl_isempty(errored_callbacks) then
+    error(string.format("Error executing 'on_keystroke' with indeces of '%s'", table.concat(errored_callbacks, ", ")))
```
```suggestion
    error(string.format("Error executing 'on_keystroke' with indices of '%s'", table.concat(errored_callbacks, ", ")))
```

prabirshrestha commented:
Why not create a vim function that can be used in both Vim script  and lua?

jamessan reviewed (commented):
I agree that it's cool how easy this implementation was, but the functionality seems a bit niche to me.

kdheepak commented:
@tjdevries Is this only enabled when using the `-W` or `-w` flag? Or does it work even if you start nvim normally?

tjdevries commented on runtime/doc/options.txt:
```diff
@@ -3504,6 +3504,12 @@ A jump table for the options with a short description can be found at |Q_op|.
 	The 'keymodel' option is set by the |:behave| command.
 
+					*'keystrokecallback'* *'kscb'*
+'keystrokecallback' 'kscb'  boolean (default: false)
```
Hmm, yes, that is a good point.

The reason I want to do this (have an option to disable it) is I don't want to have to open a new lua stack and check the table every time if people don't want it.

Perhaps a buffer local option would be better? 

The reason I don't want to add a new autocmd event is that I want the keystroke callback to be as fast as possible. I don't want to parse commands, call vim functions, etc. I just want to pass a string to luajit and let it do its work.

I'm not sure if that's a good enough reason, but that's how I was feeling when I wrote it.

tjdevries replied:
Oh, and change the name to `keystrokelog` or similar? Maybe `enableonkeystroke`? I can think of the name a bit more.

tjdevries commented on runtime/doc/lua.txt:
```diff
@@ -801,6 +801,28 @@ vim.region({bufnr}, {pos1}, {pos2}, {type}, {inclusive})       *vim.region()*
+        calling.
+
+        Parameters: ~
+            {id}: (string)  Name of the ID.
```
Yeah, that seems better. I'll change it to a namespace. That way it will make more sense to clear as well, since you're clearing from the namespace of your app.

tjdevries commented:
It always works @kdheepak no need for -w/-W flags

tjdevries commented:
> Why not create a vim function that can be used in both Vim script and lua?

To be honest, I did't want to handle both here for maximum performance. If it is going to be called on every keystroke, I want it to be as lightning fast as possible, and Lua is a first-class language for Neovim now, so every feature does not always have to have a corresponding vim function (at least in my opinion).

I suppose we could do something like

```vim
function keystroke#register(name, f)
  call luaeval(vim.register_keystroke_callback(_A.name, _A.f)", { 'name': a:name, 'f': a:f })
endfunction
```

but I want it to be all Lua side first.

jamessan commented:
Have you measured the performance hit of jumping to Lua to check whether anything's registered rather than checking an option?

Setting a global option just seems icky to me. :) I can certainly understand not wanting to make this an autocmd.  I mention it mainly because I think this would be _more_ useful if it's not (solely) a global callback.  I think there's going to be a need to either know at registration time that the callback is limited to a specific context (buffer or window) or for the callback to be provided that context.

Vim's [`popup_create()`](https://vimhelp.org/popup.txt.html#popup_create-arguments) has a 'filter' option that [specifies a callback](https://vimhelp.org/popup.txt.html#popup-filter) allowing the popup to choose which keys have an effect and what they do.  This PR could help lay some of the groundwork for the popup sugar on our floating windows.

tjdevries commented:
Ah, thanks @jamessan I will check that out and then maybe grab some of the code required to make that feature work later as well. I haven't actually measured anything in terms of the performance, I just wanted to make it as lean as possible because it happens every keystroke and so far no one has wanted to do this in vimscript or others before, so I didn't think people would want to do it now either ;)

I mean, when I have it enabled and it's checking stuff and adding to tables and all that, I cannot tell any difference between on or off, so I guess it's not terrible to just leave it on all the time -- I was just hoping to make it so people can literally just "remove" this feature from existing by disabling a flag.

@jamessan do you think having a `register_buf_keystroke_callback` and `register_keystroke_callback` as two separate functions would be good? Or should I do an options dict with `buffer` as one of the opts, to only attach to a certain buffer.

teto commented:
I agree with @jamessan this seems pretty niche, what's the usecase ? When I watch vimcasts, keys can be intercepted and shown by the terminal recorder I believe

tjdevries commented:
> I agree with @jamessan this seems pretty niche, what's the usecase ? When I watch vimcasts, keys can be intercepted and shown by the terminal recorder I believe

For example:
1. You could have a screen logger that does not log your key strokes in your terminal buffers, so that you don't display any keystrokes for password / credentials
2. You could log for only certain filetypes / modes / etc.
3. You could make a plugin for checking if you have used the same keystroke / command several times in a row, and let you know that you should be using some other movement.

Probably more, but those are just examples that I know of.  Additionally, people clearly already want to be able to track the motions and that's why people have passed around the patch that was mentioned in @kdheepak 's links. I'm proposing a better way than watching a file, but instead actually being able to directly run what you want on a per-keystroke basis.

It seems completely non-intrusive since it can be disabled by an option (or perhaps not, but I think an option is a good way to make it off-by-default).

I think it is a worthwhile addition 😄 

I will make updates to try and accommodate the feedback above though.

prabirshrestha commented:
@tjdevries my thought was to use `vim.fn.keystroke_listener_add`, `vim.fn.keystroke_listener_remove` which works in vimscript then interfcept in the lua in future to improve perf by `vim.fn.keystroke_listener_add = vim._native_lua_keystroke_listener_add`.

tjdevries commented:
Yes, that is a good idea.

norcalli commented:
I proposed something similar regarding this and Justin, Bjorn, and I discussed that it was also related to user created modes (like alternatives to normal,terminal,insert,etc.). I would say that merging this might be premature considering there is more to the design that could be considered, like modiying the keystrokes, per buffer callbacks. So I'm going to :-1: for now.

tjdevries commented:
Sorry, I forgot to move it back to WIP after the commetns.

Yeah, I will go back and answer the questions above, but I think it would be very simple to answer a few of the questions above (like per-buffer callbacks, for example).

I don't know why you'd need to modify the keystrokes for this? This is just to see what people were typing.

I see this PR as a natural extension of the `-w` and `-W` flags, which are useful, but not for any real-time feedback. This extends that feature in a more powerful and extensible way. Of course, since the callbacks are in Lua, people can do whatever they want with it afterward.

tjdevries commented:
@norcalli @jamessan 

I have removed the keystroke handling part of this PR, as we'll need to have a bit more discussion about how that will work (especially with forward thinking about user modes and more). This will be handled in a separate PR.

 I have removed the global option, as it is pretty useless and icky (as mentioned by @jamessan ). 

I have kept the tests that are applicable and updated the documentation as well.

While I understand the PR seems somewhat niche, there are already plugins and people doing interesting things based off of this WIP PR (see https://github.com/theprimeagen/vim-apm and https://github.com/ThePrimeagen/vim-apm/tree/timings) and I'd like to add some of my own features on top of it as well.

It is not possible to reproduce this kind of work just from the output of -W since you cannot run code at the moment of the keystroke.

For example, to answer @jamessan 's point about "buffer local" versions of this, you can simply do something like checking if the current buffer is the one that you want to be executing (and this is how I plan on not showing keystrokes done in the terminal for my eventual screen keys

tjdevries commented:
I will not add a vimscript version of this, although if someone wanted to go through the work in the future, I will not oppose the PR.

tjdevries committed:
fixup: Forgot to remove mention of old option

jamessan commented on test/functional/lua/vim_spec.lua:
```diff
@@ -1068,6 +1068,115 @@ describe('lua stdlib', function()
+
+      helpers.insert("hello")
+
+      eq({'i', 'w', 'o', 'r', 'l', 'd', '<ESC>'}, exec_lua [[
```
How come this assertion doesn't check against the result of `table.concat(KeysPressed, '')` like the others?

jamessan commented on src/nvim/lua/executor.c:
```diff
@@ -1465,3 +1465,40 @@ void nlua_free_typval_dict(dict_T *const d)
+
+  lua_State *const lstate = nlua_enter();
+
+#ifdef DEBUG
```
```suggestion
#ifndef NDEBUG
```

jamessan commented on src/nvim/lua/executor.c:
```diff
@@ -1465,3 +1465,40 @@ void nlua_free_typval_dict(dict_T *const d)
+  // [ vim ]
+  lua_pop(lstate, 1);
+
+#ifdef DEBUG
```
```suggestion
#ifndef NDEBUG
```

jamessan commented on src/nvim/keymap.c:
```diff
@@ -530,13 +530,24 @@ unsigned int trans_special(const char_u **srcp, const size_t src_len,
+/// When "keycode" is TRUE prefer key code, e.g. K_DEL instead of DEL.
+/// The sequence is not NUL terminated.
+/// This is how characters in a string are encoded.
+unsigned int special_to_buf(int key, int modifiers, int keycode, char_u *dst)
```
```suggestion
unsigned int special_to_buf(int key, int modifiers, bool keycode, char_u *const dst)
```

tjdevries committed:
fixup: Answer jamessan comments

tjdevries commented:
Cool, answered those comments (and just removed the intermediate variable in the tests, since it didn't really need to be there at all).

norcalli commented on src/nvim/lua/executor.c:
```diff
@@ -1465,3 +1465,40 @@ void nlua_free_typval_dict(dict_T *const d)
+void nlua_execute_log_keystroke(int c)
+{
+  char_u buf[NUMBUFLEN];
+  buf[special_to_buf(c, mod_mask, false, buf)] = NUL;
```
Instead of doing this, you could do 
```suggestion
  int buf_len = special_to_buf(c, mod_mask, false, buf);
```
and then further down use `lua_pushlstring(lstate, buf, buf_len)` which precalculates the buffer length.

norcalli commented on runtime/doc/lua.txt:
```diff
@@ -876,6 +876,32 @@ vim.region({bufnr}, {pos1}, {pos2}, {type}, {inclusive})       *vim.region()*
         of linewise selections of the form `{linenr = {startcol, endcol}}` .
 
+                                             *vim.register_keystroke_callback()*
+vim.register_keystroke_callback({ns_id}, {fn})
```
Using the terminology `ns_id` implies that `vim.api.nvim_buf_clear_namespace(bufnr, ns_id)` would clear the keystroke callbacks. Otherwise using ns_id here is incorrect, I think.

```
                Clears namespaced objects (highlights, extmarks, virtual text)
                from a region.
```

Instead with what you want, I would just take `{fn}` and return an `id` which you can use later to remove the callback. This can just be an incrementing integer since someone probably won't make 4 million callbacks.

norcalli commented on runtime/doc/lua.txt:
```diff
@@ -876,6 +876,32 @@ vim.region({bufnr}, {pos1}, {pos2}, {type}, {inclusive})       *vim.region()*
+        Register a lua {fn} with an {ns_id} to be run after every keystroke.
+
+        NOTE: {fn} will be automatically removed if an error occurs while
+        calling.
```
It would be tremendously getting messages for every keystroke if you made a bug, and people could wrap it in pcall, I suppose. I'm ok with this, but I would consider adding a sentence about the motivation (but not required).

norcalli commented on src/nvim/lua/vim.lua:
```diff
@@ -489,4 +489,73 @@ function vim.defer_fn(fn, timeout)
+
+--- Function that executes the keystroke callbacks.
+--@private
+function vim._log_keystroke(char)
```
Alternatively
```lua
function vim._log_keystroke(char)
  local failed_indices = {}
  local failed_messages = {}
  for k, v in pairs(on_keystroke_callbacks) do
    local ok, err_msg = pcall(v, char)
    if not ok then
      on_keystroke_callbacks[k] = nil
      table.insert(failed_indices, k)
      table.insert(failed_messages, err_msg)
    end
  end

  if failed_indices[1] then
    error(string.format(
      "Error executing 'on_keystroke' with indices of '%s'\n    With messages: %s",
      table.concat(failed_indices, ", "),
      table.concat(failed_messages, "\n")))
  end
end
```
Note that deleting from a table is allowed during iteration with pairs. Quote:

> The behavior of next is undefined if, during the traversal, you assign any value to a non-existent field in the table. You may however modify existing fields. In particular, you may clear existing fields.

norcalli reviewed (changes requested):
Some questions and comments. Some parts we talked about in gitter but they're all also duplicated here.

norcalli commented on src/nvim/lua/vim.lua:
```diff
@@ -489,4 +489,73 @@ function vim.defer_fn(fn, timeout)
+--                    See |i_CTRL-V|
+function vim.register_keystroke_callback(ns_id, fn)
+  vim.validate {
+    fn = { fn, 'c', true},
```
If this is optional to allow using register to *also* remove things, then we should remove the remove_keystroke_callback function and say that you can use `nil` to clear it. Otherwise the optionality should be documented.

tjdevries committed:
fixup: Answer norcalli comments

tjdevries committed:
fixup: portability

Commit:
Update runtime/doc/lua.txt

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -876,6 +876,34 @@ vim.region({bufnr}, {pos1}, {pos2}, {type}, {inclusive})       *vim.region()*
 
+                                             *vim.register_keystroke_callback()*
+vim.register_keystroke_callback({fn}, {ns_id})
+        Register a lua {fn} with an {ns_id} to be run after every keystroke.
```
`-w` was mentioned in the PR discussion, that's a hint that it should be mentioned in the docs here (and the commit message).

zeertzjq commented on runtime/doc/lua.txt:
```diff
@@ -876,6 +876,34 @@ vim.region({bufnr}, {pos1}, {pos2}, {type}, {inclusive})       *vim.region()*
+            {fn}: (function): Function to call on keystroke.
+                    It should take one argument, which is a string.
+                    The string will contain the literal keys typed.
+                    See |i_CTRL-V|
```
What is the `i_CTRL-V` tag supposed to mean here? Mentioning it here is quite confusing.

justinmk replied:
It is a way to see the "literal keys typed".
